src/kernels/kernelStartupCodeProvider.node.ts

```diff
@@ -0,0 +1,80 @@
+
+import { inject, injectable } from 'inversify';
+import { Uri } from 'vscode';
+import { IWorkspaceService } from '../platform/common/application/types';
+import { CodeSnippets } from '../platform/common/constants';
+import { getFilePath } from '../platform/common/platform/fs-paths';
+import { IFileSystem } from '../platform/common/platform/types';
+import { IConfigurationService } from '../platform/common/types';
+import { calculateWorkingDirectory } from '../platform/common/utils.node';
+import { traceInfo } from '../platform/logging';
+import { isLocalHostConnection, isPythonKernelConnection } from './helpers';
+import { expandWorkingDir } from './jupyter/jupyterUtils';
+import { IKernel, isLocalConnection, IStartupCodeProvider, StartupCodePriority } from './types';
+
+@injectable()
+export class KernelStartupCodeProvider implements IStartupCodeProvider {
+    public priority = StartupCodePriority.Base;
+
+    constructor(
+        @inject(IConfigurationService) private readonly configService: IConfigurationService,
+        @inject(IFileSystem) private readonly fs: IFileSystem,
+        @inject(IWorkspaceService) private readonly workspaceService: IWorkspaceService
+    ) {}
+
+    async getCode(kernel: IKernel): Promise<string[]> {
+        if (
+            !(
+                isLocalConnection(kernel.kernelConnectionMetadata) &&
+                !this.configService.getSettings(undefined).forceIPyKernelDebugger
+            )
+        ) {
+            return [];
+        }
+
+        if (
+            isLocalConnection(kernel.kernelConnectionMetadata) ||
+            isLocalHostConnection(kernel.kernelConnectionMetadata)
+        ) {
+            let suggestedDir = await calculateWorkingDirectory(
+                this.configService,
+                this.workspaceService,
+                this.fs,
+                kernel.resourceUri
+            );
+            if (suggestedDir && (await this.fs.exists(suggestedDir))) {
+                traceInfo('UpdateWorkingDirectoryAndPath in Kernel');
+                // We should use the launch info directory. It trumps the possible dir
+                return this.getChangeDirectoryCode(kernel, suggestedDir);
+            } else if (kernel.resourceUri && (await this.fs.exists(kernel.resourceUri))) {
+                // Combine the working directory with this file if possible.
+                suggestedDir = Uri.file(
+                    expandWorkingDir(
+                        getFilePath(suggestedDir),
+                        kernel.resourceUri,
+                        this.workspaceService,
+                        this.configService.getSettings(kernel.resourceUri)
+                    )
+                );
+                if (suggestedDir && (await this.fs.exists(suggestedDir))) {
+                    traceInfo('UpdateWorkingDirectoryAndPath in Kernel');
+                    return this.getChangeDirectoryCode(kernel, suggestedDir);
+                }
+            }
```

Comment:
I'm not suer how much it would make sense to merge the tow ways to resolve the suggested dir, but I feel that with a new local method the code can be much simpler to parse. Something like:

```ts
private async resolveDirectory(kernel: IKernel, path?: Uri): Promise<Uri | undefined> {
    let dir: Uri | undefined;
    if (!path) {
        dir = await calculateWorkingDirectory(
            this.configService,
            this.workspaceService,
            this.fs,
            kernel.resourceUri
        )
    } else {
        dir = Uri.file(
            expandWorkingDir(
                getFilePath(path),
                kernel.resourceUri,
                this.workspaceService,
                this.configService.getSettings(kernel.resourceUri)
            )
        );
    }
    if (dir && (await this.fs.exists(dir))) {
        return dir;
    }
    return;
}
```

Would allow us to do something like:

```ts
        if (
             isLocalConnection(kernel.kernelConnectionMetadata) ||
             isLocalHostConnection(kernel.kernelConnectionMetadata)
         ) {
             let suggestedDir = await this.resolveDirectory(kernel);
             if (kernel.resourceUri && (await this.fs.exists(kernel.resourceUri))) {
                 suggestedDir = await this.resolveDirectory(kernel, suggestedDir);
             }
             if (suggestedDir) {
                 traceInfo('UpdateWorkingDirectoryAndPath in Kernel');
                 return this.getChangeDirectoryCode(kernel, suggestedDir);
             }
         }
```

Reply:
if this doesn't look valuable, feel free to resolve it!

Reply:
Thanks, I'll consider having a new PR taking this suggestion!